File: java/leetcode/backtracking/permutations.java
/**
 * 46. Permutations
 * Given a collection of distinct integers, return all possible permutations.

Example:

Input: [1,2,3]
Output:
[
  [1,2,3],
  [1,3,2],
  [2,1,3],
  [2,3,1],
  [3,1,2],
  [3,2,1]
]
 */

import java.util.List;
import java.util.ArrayList;

public class permutations {

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> permutes = new ArrayList<>();
        List<Integer> permuteList = new ArrayList<>();
        boolean[] hasVisited = new boolean[nums.length];
        backtracking(permuteList, permutes, hasVisited, nums);
        return permutes;
    }

    private void backtracking(List<Integer> permuteList, List<List<Integer>> permutes, 
        boolean[] visited, final int[] nums) {
        
        if (permuteList.size() == nums.length) {
            permutes.add(new ArrayList<>(permuteList));
            return;
        }
        for (int i = 0; i < visited.length; i++) {
            if (visited[i]) {
                continue;
            }
            visited[i] = true;
            permuteList.add(nums[i]);
            backtracking(permuteList, permutes, visited, nums);
            permuteList.remove(permuteList.size()-1);
            visited[i] = false;
        }
    }
}

File: golang/leetcode/backtracking/permutations.go
package backtracking

// 46. Permutations

func permute(nums []int) [][]int {
	permutes := make([][]int, 0)
	temp := make([]int, 0)
	visited := make([]bool, len(nums))
	permuteTrack(temp, &permutes, visited, nums)
	return permutes
}

func permuteTrack(temp []int, permutes *[][]int, visited []bool, nums []int) {
	if len(temp) == len(nums) {
		combo := make([]int, len(temp))
		copy(combo, temp)
		*permutes = append(*permutes, combo)
		return
	}
	for i := 0; i < len(visited); i++ {
		if visited[i] {
			continue
		}
		visited[i] = true
		temp = append(temp, nums[i])
		permuteTrack(temp, permutes, visited, nums)
		temp = temp[:len(temp)-1]
		visited[i] = false
	}
}
